src/uct/ib/rc/accel/rc_mlx5_iface.c

```diff
@@ -170,20 +170,44 @@ static ucs_status_t uct_rc_mlx5_iface_arm_cq(uct_ib_iface_t *ib_iface,
 #endif
 }
 
+static ucs_status_t
+uct_rc_mlx5_ep_handle_failure(uct_rc_mlx5_ep_t *ep, ucs_status_t status)
+{
+    uct_ib_iface_t *ib_iface = ucs_derived_of(ep->super.super.super.iface,
+                                              uct_ib_iface_t);
+    uct_rc_iface_t *rc_iface = ucs_derived_of(ib_iface, uct_rc_iface_t);
+
+    uct_rc_txqp_purge_outstanding(&ep->super.txqp, status, 0);
+    /* poll_cqe for mlx5 returns NULL in case of failure and the cq_avaialble
+       is not updated for the error cqe and all outstanding wqes*/
+    rc_iface->tx.cq_available += ep->tx.wq.bb_max -
+                                 uct_rc_txqp_available(&ep->super.txqp);
+    return ib_iface->ops->set_ep_failed(ib_iface, &ep->super.super.super,
+                                        status);
+}
+
 static void
 uct_rc_mlx5_iface_handle_failure(uct_ib_iface_t *ib_iface, void *arg,
                                  ucs_status_t status)
 {
-    struct mlx5_cqe64 *cqe    = arg;
-    uct_rc_iface_t    *iface  = ucs_derived_of(ib_iface, uct_rc_iface_t);
-    unsigned          qp_num  = ntohl(cqe->sop_drop_qpn) & UCS_MASK(UCT_IB_QPN_ORDER);
-    uct_rc_mlx5_ep_t  *ep     = ucs_derived_of(uct_rc_iface_lookup_ep(iface, qp_num),
-                                               uct_rc_mlx5_ep_t);
-    ucs_log_level_t   log_lvl = UCS_LOG_LEVEL_FATAL;
+    struct mlx5_cqe64  *cqe;
+    uct_rc_iface_t     *iface;
+    uct_rc_mlx5_ep_t   *ep;
     uct_ib_mlx5_txwq_t txwq_copy;
-    size_t            txwq_size;
-    ucs_status_t      ep_status;
+    size_t             txwq_size;
+    unsigned           qp_num;
+    ucs_log_level_t    log_lvl;
 
+    if (status == UCS_ERR_CANCELED) {
```

Comment:
need to put the code block which handles error reporting under if (status != CANCELED), so uct_rc_mlx5_ep_handle_failure() will be called only from one place in the code
- same comment for uct_rc_verbs_handle_failure

Reply:
this will add several branches in this function and make code hard readable.

Reply:
@evgeny-leksikov on the other hand, after doing this, you will see the bug in your code (also in verbs flow)

Reply:
changed, but seems like i'm missing something, could you clarify pls?

Reply:
the code is passing wc/cqe to a function which expects ep:
```
uct_rc_mlx5_ep_handle_failure(arg, status);
```
vs
```
if (uct_rc_mlx5_ep_handle_failure(ep, status) == UCS_OK) {
```
(AFAIR there was similar bug in UD transport)